pkg/controller/volume/scheduling/scheduler_binder.go

```diff
@@ -758,27 +762,37 @@ func (b *volumeBinder) findMatchingVolumes(pod *v1.Pod, claimsToBind []*v1.Persi
 func (b *volumeBinder) checkVolumeProvisions(pod *v1.Pod, claimsToProvision []*v1.PersistentVolumeClaim, node *v1.Node) (provisionSatisfied bool, provisionedClaims []*v1.PersistentVolumeClaim, err error) {
 	podName := getPodName(pod)
 	provisionedClaims = []*v1.PersistentVolumeClaim{}
-
 	for _, claim := range claimsToProvision {
 		pvcName := getPVCName(claim)
 		className := v1helper.GetPersistentVolumeClaimClass(claim)
 		if className == "" {
+			// Returning err below adds event to the *pod*. Add similar event to the claim.
+			b.recorder.Eventf(claim, pod, v1.EventTypeWarning, "FailedScheduling", "Provisioning", "didn't find available persistent volumes and no storage class is set")
 			return false, nil, fmt.Errorf("no class for claim %q", pvcName)
 		}
 
 		class, err := b.classLister.Get(className)
 		if err != nil {
+			// Returning err below adds event to the *pod*. Add similar event to the claim.
+			msg := fmt.Sprintf("didn't find available persistent volumes and failed to get storage class %q", className)
+			b.recorder.Eventf(claim, pod, v1.EventTypeWarning, "FailedScheduling", "Provisioning", msg)
 			return false, nil, fmt.Errorf("failed to find storage class %q", className)
 		}
 		provisioner := class.Provisioner
 		if provisioner == "" || provisioner == pvutil.NotSupportedProvisioner {
 			klog.V(4).Infof("storage class %q of claim %q does not support dynamic provisioning", className, pvcName)
+			// Returning false below adds event to the *pod*. Add similar event to the claim.
+			msg := fmt.Sprintf("didn't find available persistent volumes and dynamic provisioning is disabled for storage class %q", className)
+			b.recorder.Eventf(claim, pod, v1.EventTypeWarning, "FailedScheduling", "Provisioning", msg)
 			return false, nil, nil
 		}
 
 		// Check if the node can satisfy the topology requirement in the class
 		if !v1helper.MatchTopologySelectorTerms(class.AllowedTopologies, labels.Set(node.Labels)) {
 			klog.V(4).Infof("Node %q cannot satisfy provisioning topology requirements of claim %q", node.Name, pvcName)
+			// Returning false below adds event to the *pod*. Add similar event to the claim.
+			msg := fmt.Sprintf("Node %q cannot satisfy provisioning topology requirements of required storage class", node.Name)
```

Comment:
This is going to generate an event for every node in the cluster that it fails on, which in the case of local volumes, could be all nodes - 1.

Reply:
Oops, this PR is completely wrong then. I tested only on local-up-cluster, which has a single node.